dashboard/app/controllers/projects_controller.rb

```diff
@@ -366,20 +366,33 @@ def remix
     rescue ArgumentError, OpenSSL::Cipher::CipherError
       return head :bad_request
     end
+    project_type = params[:key]
     new_channel_id = ChannelToken.create_channel(
       request.ip,
       StorageApps.new(storage_id('user')),
       src: src_channel_id,
-      type: params[:key],
+      type: project_type,
       remix_parent_id: remix_parent_id,
     )
-    AssetBucket.new.copy_files src_channel_id, new_channel_id
-    animation_list = AnimationBucket.new.copy_files src_channel_id, new_channel_id
+    AssetBucket.new.copy_files src_channel_id, new_channel_id if uses_asset_bucket?(project_type)
+    animation_list = uses_animation_bucket?(project_type) ? AnimationBucket.new.copy_files(src_channel_id, new_channel_id) : []
     SourceBucket.new.remix_source src_channel_id, new_channel_id, animation_list
-    FileBucket.new.copy_files src_channel_id, new_channel_id
+    FileBucket.new.copy_files src_channel_id, new_channel_id if uses_file_bucket?(project_type)
     redirect_to action: 'edit', channel_id: new_channel_id
   end
 
+  private def uses_asset_bucket?(project_type)
+    %w(applab makerlab gamelab spritelab).include? project_type
+  end
+
+  private def uses_animation_bucket?(project_type)
+    %w(gamelab spritelab).include? project_type
+  end
+
+  private def uses_file_bucket?(project_type)
+    %w(weblab).include? project_type
+  end
```

Comment:
What happens if a bucket that's supposed to be used by a project type fails to get copied over? Will it fail obviously in that the things in that bucket won't be available in the remixed version, or fail subtly in that the remixed version will still be referencing the things in the original bucket?

If the latter, I'd be worried about a scenario in which we add a new project type and forget to go in here and update these methods appropriately. If the former, I'm not as worried about it because I think we'll catch it.

Reply:
I'm fairly worried about this too.  In general we keep references project-local, but I think it depends somewhat on the implementation of the project type and it's possible we've got some subtle vulnerability here.

- **App Lab** has no manifest for assets - it LISTs them to populate the assets dialog, and references to asset names in student code are assumed to be relative to the local project.  If we forgot to copy the assets bucket, assets referenced in student code would simply be unavailable in the remixed project and it would be obviously broken.
- **Web Lab** uses a server-managed manifest for files (`manifest.json`) also stored in the files bucket that looks like this:
  ```json
  [
    {
      "filename":"index.html",
      "category":"text",
      "size":4551,
      "versionId":"qeiQ9v2TbAoQsGkc_.5KWtmmes3YTezX"
     }, 
    {
      "filename":"style.css",
      "category":"text",
      "size":79,
      "versionId":"8oo3e9LeaK5a9uRCpXHkDH_AeUSK53jP"
    }
  ]
  ```
  Again, filenames are relative to the project... and we wouldn't copy the manifest at all if we forgot this bucket, so I think this would also be obviously broken. [Here's the remix logic for Files/Web Lab if you're interested.](https://github.com/code-dot-org/code-dot-org/blob/1596e89cefd8f807cf9b98c2ae1a2710aca257a9/shared/middleware/helpers/file_bucket.rb#L25-L43)
  Web Lab stores the version id of `manifest.json` into its sources `main.json` to interact nicely with our version history system.  It still looks for `manifest.json` under the new storage/channel id so this would also break in an obvious way if we forgot to remix files.
- **Game Lab** works the same as App Lab with regard to assets.   Animations might be a different story - Game Lab stores its animation manifest in the sources bucket, and [updates that manifest on remix](https://github.com/code-dot-org/code-dot-org/blob/1596e89cefd8f807cf9b98c2ae1a2710aca257a9/shared/middleware/helpers/source_bucket.rb#L73-L111) using the results of the animation bucket copy operation.  This step is skipped if no animation bucket copy occurs.  I believe what you'd end up with is an outdated manifest, still looking in the local project for assets that weren't copied, and it should be obviously broken.  However I know we explored cross-project animation references in the past as a way of minimizing storage of duplicate animations.  We do this for library animations (they are absolute URLs) but abandoned this strategy on remix because of the complexity it introduced.  I'll actually try this on my local machine and see what happens.

Reply:
Initial project's `main.json` before remix (where `smiley_monitor` is a local project animation and `orange_1` is a library animation) :
```json
{
  "source":"var spritea = createSprite(100, 200);\nspritea.setAnimation(\"smiley_monitor\");\nvar spriteb = createSprite(300, 200);\nspriteb.setAnimation(\"orange_1\");\nfunction draw() {\n  background(rgb(200, 200, 200));\n  drawSprites();\n}\n",
  "animations":{
    "orderedKeys":[
      "c6d6d275-8108-4610-bbe0-8c211553f5f3",
      "b27090ab-03d5-47ef-8fc8-7f01e1a2de14"
    ],
    "propsByKey":{
      "c6d6d275-8108-4610-bbe0-8c211553f5f3":{
        "name":"smiley_monitor",
        "sourceUrl":null,
        "frameSize":{
          "x":136,
          "y":129
        },
        "frameCount":1,
        "looping":true,
        "frameDelay":12,
        "version":"kTY7ZZdXpZ0An027nlaxHl8GvwdU.TAB"
      },
      "b27090ab-03d5-47ef-8fc8-7f01e1a2de14":{
        "name":"orange_1",
        "sourceUrl":"/api/v1/animation-library/HOm1OJiR.3NomTzDwN2hXRFIkstT2CbB/category_food/orange.png",
        "frameSize":{
          "x":300,
          "y":296
        },
        "frameCount":1,
        "looping":true,
        "frameDelay":2,
        "version":"HOm1OJiR.3NomTzDwN2hXRFIkstT2CbB"
      }
    }
  }
}
```

I disabled the animation bucket copy locally and remixed the project.  I confirmed that the resulting `main.json` is identical to the above, so the library animation is an absolute reference and still good, but the local animation is now an invalid reference within the local project.  Sure enough, on remix I immediately get this error:

![image](https://user-images.githubusercontent.com/1615761/47171324-48648800-d2bd-11e8-986f-b95444f000de.png)

Reply:
Thanks for the thorough investigation! I feel much more confident in this change now